pricegraph/src/graph/bellman_ford.rs

```diff
@@ -0,0 +1,149 @@
+//! The module contains a slightly modified version of the `petgraph`
+//! implementation of the Bellman-Ford graph search algorigthm that returns the
+//! detected negative cycle on error.
+
+use petgraph::algo::FloatMeasure;
+use petgraph::visit::{
+    Data, EdgeRef, GraphBase, IntoEdges, IntoNodeIdentifiers, NodeCount, NodeIndexable,
+};
+
+/// A type definition for the result of a Bellman-Ford shortest path search
+/// containing the weight of the shortest paths and the predecessor vector for
+/// the shortests paths.
+pub type Paths<G> = (
+    Vec<<G as Data>::EdgeWeight>,
+    Vec<Option<<G as GraphBase>::NodeId>>,
+);
+
+/// A negative cycle error with a path representing the cycle.
+pub struct NegativeCycle<G: Data>(pub Vec<G::NodeId>);
```

Comment:
Would it make sense to also expose the weight as it relates to the surplus that can be extracted from the cycle?

Reply:
We can add it if we end up using it. We actually need to walk to cycle to fill it (since we need to find the capacity of the cycle - i.e. the minimum amount that can be sent around) and for that we use the prices and not the log of the prices to do the calculations.

Reply:
So for now we don't use it and it would add extra complexity to the `find_cycle` so I imagine we can leave it out for now.